Should no-redeclare also check functions?

I just got burned by a stupid mistake, but I think it should've been caught by `no-redeclare`.

Given the following code:

```
var a = true;
function a() {}
```

If I turn on `no-redeclare: 2`, this is **not** treated as an error, but I think it should.

To be clear, this was a stupid mistake on my part. I added a `var` to the top of a file that was named the same as a function declared much further down in my code. Nonetheless, I think `no-redeclare` should take hoisted functions into account as well.

I tried:

```
> eslint -v
v0.24.0
> echo "var a = true; function a() {}" | eslint --stdin --env browser --env es6 --rule "no-redeclare:2" --reset --no-eslintrc

<text>
  1:24  error  a is already defined  no-redeclare

✖ 1 problem (1 error, 0 warnings)
```

If you have turned `ecmaFeatures: {modules: true}` on, yes, the `no-redeclare` rule had been missing global declarations. And it was fixed on `master` branch now.

Related: #2903

--- a/tests/lib/rules/no-redeclare.js
+++ b/tests/lib/rules/no-redeclare.js
@@ -33,6 +33,8 @@ eslintTester.addRuleTest("lib/rules/no-redeclare", {
         { code: "switch(foo) { case a: let b = 3;\ncase b: let b = 4}", ecmaFeatures: { blockBindings: true }, errors: [{ message: "b is already defined", type: "Identifier"}] },
         { code: "var a = 3; var a = 10;", errors: [{ message: "a is already defined", type: "Identifier"}] },
         { code: "var a = {}; var a = [];", errors: [{ message: "a is already defined", type: "Identifier"}] },
+        { code: "var a; function a() {}", errors: [{ message: "a is already defined", type: "Identifier" }] },
+        { code: "function a() {} function a() {}", errors: [{ message: "a is already defined", type: "Identifier" }] },
         { code: "var a = function() { }; var a = function() { }", errors: [{ message: "a is already defined", type: "Identifier"}] },
         { code: "var a = function() { }; var a = new Date();", errors: [{ message: "a is already defined", type: "Identifier"}] },
         { code: "var a = 3; var a = 10; var a = 15;", errors: [{ message: "a is already defined", type: "Identifier"}, { message: "a is already defined", type: "Identifier"}] },

